bot/utils/redis_cache.py

```diff
@@ -0,0 +1,215 @@
+
+    The class this object is instantiated in must also contains an attribute with an
+    instance of Bot. This is because Bot contains our redis_pool, which is how this
+    class communicates with the Redis server.
+
+    We implement several convenient methods that are fairly similar to have a dict
+    behaves, and should be familiar to Python users. The biggest difference is that
+    all the public methods in this class are coroutines, and must be awaited.
+
+    Because of limitations in Redis, this cache will only accept strings, integers and
+    floats both for keys and values.
+
+    Simple example for how to use this:
+
+    class SomeCog(Cog):
+        # To initialize a valid RedisCache, just add it as a class attribute here.
+        # Do not add it to the __init__ method or anywhere else, it MUST be a class
+        # attribute. Do not pass any parameters.
+        cache = RedisCache()
+
+        async def my_method(self):
+            # Now we can store some stuff in the cache just by doing this.
+            # This data will persist through restarts!
+            await self.cache.set("key", "value")
+
+            # To get the data, simply do this.
+            value = await self.cache.get("key")
+
+            # Other methods work more or less like a dictionary.
+            # Checking if something is in the cache
+            await self.cache.contains("key")
+
+            # iterating the cache
+            async for key, value in self.cache.items():
+                print(value)
+
+            # We can even iterate in a comprehension!
+            consumed = [value async for key, value in self.cache.items()]
+    """
+
+    _namespaces = []
+
+    def __init__(self) -> None:
+        """Raise a NotImplementedError if `__set_name__` hasn't been run."""
+        self._namespace = None
+        self.bot = None
+
+    def _set_namespace(self, namespace: str) -> None:
+        """Try to set the namespace, but do not permit collisions."""
+        while namespace in self._namespaces:
+            namespace += "_"
+
+        self._namespaces.append(namespace)
+        self._namespace = namespace
+
+    @staticmethod
+    def _to_typestring(value: ValidRedisType) -> str:
+        """Turn a valid Redis type into a typestring."""
+        if isinstance(value, float):
+            return f"f|{value}"
+        elif isinstance(value, int):
+            return f"i|{value}"
+        elif isinstance(value, str):
+            return f"s|{value}"
+
+    @staticmethod
+    def _from_typestring(value: Union[bytes, str]) -> ValidRedisType:
+        """Turn a typestring into a valid Redis type."""
+        if isinstance(value, bytes):
+            value = value.decode('utf-8')
+        if value.startswith("f|"):
+            return float(value[2:])
+        if value.startswith("i|"):
+            return int(value[2:])
+        if value.startswith("s|"):
+            return value[2:]
```

Comment:
I wonder if these could be defined in one place?

A very simple example:
```python
TEMPLATES = (
    (float, "f|{value}"),
    (int,   "i|{value}"),
    (str,   "s|{value}"),
)
```
Both functions then use this definition, e.g. by lopping through the pairs until it finds a match. I don't love that this is essentially a bi-directional map with each direction defined elsewhere, seems fragile. Both functions also return `None` should no match be found, which is not annotated nor documented. I looked through the entire cog, so maybe that cannot happen - these are just thoughts that I'm having now.

Reply:
An interesting thought. I'll make an attempt at this.

Reply:
Perhaps better to define the prefix, not the template, e.g.:
```python
(float, "f|")
```
Makes it easier for the `startswith` check.

Reply:
Yep, that's what I went for. It's addressed in 5120717

Reply:
Solid work - thanks for accepting the suggestion. I think this ends up being a lot safer and convenient, as the `TYPESTRING_PREFIXES` definition now also serves as documentation.